### PR TITLE
Allow Organisers and Delegates to register for competitions before they open

### PR DESCRIPTION
**Pull request by simonkellly**
This lets listed competition staff register before the normal registration period opens, solving issue #7073.

### CODE REVIEW

**Comment by simonkellly** [comment]
I am not 100% sure what to do in terms of tests for this (I probably should have opened this as a draft)

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 1]
```diff
@@ -604,7 +604,7 @@ def refund_payment
 
   def create
     @competition = competition_from_params
-    if !@competition.registration_opened?
+    if !@competition.registration_opened? && !@competition.user_can_pre_register?(current_user)
```
Comment (gregorbg) [inline WcaOnRails/app/controllers/registrations_controller.rb:607]: Ruby nit: `if !foo` should become `unless foo`. To be fair, this was already wrong before your commit, but now that you're touching this line might as well clean it up :upside_down_face: 

`unless @competition.registration_opened? || @competition.user_can_pre_register?(current_user)`

File: WcaOnRails/app/views/registrations/register.html.erb [hunk 2]
```diff
@@ -32,7 +32,8 @@
     <% end %>
   <% else %>
     <% user_may_register = current_user.cannot_register_for_competition_reasons.length == 0 %>
-    <% if @registration.show_details? %>
+    <% current_user_can_pre_register = @competition.user_can_pre_register?(current_user) %>
+    <% if @registration.show_details? || current_user_can_pre_register %>
```
Comment (gregorbg) [inline WcaOnRails/app/views/registrations/register.html.erb:36]: Why do you need this additional `OR` check here?
If it is absolutely necessary, can you perhaps merge it into the `show_details?` method? Or will that have too many side effects?

File: WcaOnRails/app/views/registrations/register.html.erb [hunk 3]
```diff
@@ -82,7 +83,7 @@
         </div>
       <% end %>
 
-      <% if @registration.show_details? %>
+      <% if @registration.show_details? || current_user_can_pre_register%>
```
Comment (gregorbg) [inline WcaOnRails/app/views/registrations/register.html.erb:86]: See above about the additonal `OR` check.

**Review by gregorbg — CHANGES_REQUESTED** [review]
Thanks for the contribution! What is troubling you about the tests in particular?

File: WcaOnRails/app/controllers/registrations_controller.rb (hunk 1 above)
Comment (simonkellly) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Nifty, I have made this change

File: WcaOnRails/app/views/registrations/register.html.erb (hunk 2 above)
Comment (simonkellly) [reply WcaOnRails/app/views/registrations/register.html.erb]: Yeah, the show_details method is only really used for this purpose anyway so I have gone ahead and merged them.

**Comment by simonkellly** [comment]
I was not sure if this is the kind of thing that tests should be written for or not

**Comment by gregorbg** [comment]
We most certainly want tests for this, to make sure that your "Average Joe" competitor cannot pre-register. Please consider the following test cases:

1. A Delegate pre-registers. This test should expect to work / return true.
2. A random competitor pre-registers. This test should expect to fail / return false.
3. A Delegate registers during the normal, public registration phase. This should work.

Also see if you can find existing tests, I'd bet there are some that check existing behavior mostly for Average Joe competitors.

**Comment by simonkellly** [comment]
I have added tests which cover the logic for a delegate pre registering as well as for the not already handled behaviour for organisers and registered speedcubers. I don't have much experience in writing tests so I ended up copying much of the logic from the other tests so there is a strong chance I missed something.

**Review by gregorbg — APPROVED** [review]
The tests are fine! Thanks for figuring that one out so easily by yourself :D

One nitpick: When you expect any text field or flash warning to contain localized text, you should refer to `I18n.t("foo.bar.key")` directly in case we decide to change the content of the text in the source locale at some point.
But I understand that this behavior might be copy/pasted from existing tests, so I won't blame you for that. Otherwise LGTM!

**Pull request merged**